sdk/samples/iot/hub/src/paho_iot_hub_pnp_example.c

```diff
@@ -0,0 +1,1020 @@
+
+  return AZ_OK;
+}
+// Send the twin reported property to the service
+static int send_reported_temperature_property(double temp_value, int32_t version)
+{
+  int rc;
+  printf("Sending reported property\n");
+
+  // Get the topic used to send a reported property update
+  az_span request_id_span = get_request_id();
+  if (az_failed(
+          rc = az_iot_hub_client_twin_patch_get_publish_topic(
+              &client,
+              request_id_span,
+              reported_property_topic,
+              sizeof(reported_property_topic),
+              NULL)))
+  {
+    printf("Unable to get twin document publish topic, return code %d\n", rc);
+    return rc;
+  }
+
+  // Twin reported properties must be in JSON format. The payload is constructed here.
+  az_json_builder json_builder;
+  if (version < 0)
+  {
+    if (az_failed(
+            rc = build_reported_property(
+                &json_builder, AZ_SPAN_FROM_STR("maxTempSinceLastReboot"), temp_value)))
+    {
+      return rc;
+    }
+  }
+  else
+  {
+    if (az_failed(
+            rc = build_confirmed_reported_property(
+                &json_builder, desired_temp_property_name, temp_value, 200, version, AZ_SPAN_FROM_STR("success"))))
+    {
+      return rc;
+    }
+  }
+  az_span json_payload = az_json_builder_get_json(&json_builder);
+
+  printf("Payload: %.*s\n", az_span_size(json_payload), (char*)az_span_ptr(json_payload));
+
+  // Publish the reported property payload to IoT Hub
+  rc = mqtt_publish_message(reported_property_topic, json_payload, 0);
+
+  return rc;
+}
+
+// Parse the desired temperature property from the incoming JSON
+static az_result parse_twin_desired_temperature_property(
+    az_span twin_payload_span,
+    bool is_twin_get,
+    uint32_t* parsed_value,
+    int32_t* version_number)
+{
+  az_json_parser jp;
+  bool desired_found = false;
+  AZ_RETURN_IF_FAILED(az_json_parser_init(&jp, twin_payload_span, NULL));
+  AZ_RETURN_IF_FAILED(az_json_parser_next_token(&jp));
+  if (jp.token.kind != AZ_JSON_TOKEN_BEGIN_OBJECT)
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  if (is_twin_get)
+  {
+    // If is twin get payload, we have to parse one level deeper for "desired" wrapper
+    AZ_RETURN_IF_FAILED(az_json_parser_next_token(&jp));
+    while (jp.token.kind != AZ_JSON_TOKEN_END_OBJECT)
+    {
+      if (az_json_token_is_text_equal(&jp.token, desired_property_name))
+      {
+        desired_found = true;
+        AZ_RETURN_IF_FAILED(az_json_parser_next_token(&jp));
+        break;
+      }
+      else
+      {
+        // else ignore token.
+        AZ_RETURN_IF_FAILED(az_json_parser_skip_children(&jp));
+      }
+
+      AZ_RETURN_IF_FAILED(az_json_parser_next_token(&jp));
+    }
+  }
+  else
+  {
+    desired_found = true;
+  }
+
+  if (!desired_found)
+  {
+    printf("Desired property object not found in twin");
+    return AZ_ERROR_ITEM_NOT_FOUND;
+  }
+
+  if (jp.token.kind != AZ_JSON_TOKEN_BEGIN_OBJECT)
```

Comment:
Should we move the parser by one more token before checking this? If `is_twin_get` is true, and we go through the loop above, find `desired_found`, and then exit the loop when `jp.token.kind == AZ_JSON_TOKEN_END_OBJECT), won't this check fail and we return `AZ_ERROR_PARSER_UNEXPECTED_CHAR`?

If that is actually the desired behavior, ignore my question.